File: ios/Classes/FileCryptor.swift
import Foundation

/// Streams a file through a CipherAdapter off the main thread in 1 MB chunks.
/// Encrypt == decrypt for all built-in schemes (XOR keystreams).
final class FileCryptor {

    struct Progress {
        let operationId: String
        let bytesProcessed: Int64
        let totalBytes: Int64
        let done: Bool
        var error: String?
        var errorCode: String?

        var asMap: [String: Any] {
            var m: [String: Any] = [
                "operationId": operationId,
                "bytesProcessed": bytesProcessed,
                "totalBytes": totalBytes,
                "done": done,
            ]
            if let error { m["error"] = error }
            if let errorCode { m["errorCode"] = errorCode }
            return m
        }
    }

    static let shared = FileCryptor()
    private let queue = DispatchQueue(label: "svp.file-cryptor", qos: .utility)
    private var cancelled = Set<String>()
    private let lock = NSLock()

    private let chunk = 1024 * 1024

    func start(
        inputPath: String,
        outputPath: String,
        adapter: CipherAdapter,
        onProgress: @escaping (Progress) -> Void
    ) -> String {
        let id = UUID().uuidString
        queue.async { [self] in
            guard let input = FileHandle(forReadingAtPath: inputPath) else {
                onProgress(Progress(operationId: id, bytesProcessed: 0, totalBytes: 0,
                                    done: true, error: "Input not found: \(inputPath)",
                                    errorCode: "fileNotFound"))
                return
            }
            defer { try? input.close() }

            FileManager.default.createFile(atPath: outputPath, contents: nil)
            guard let output = FileHandle(forWritingAtPath: outputPath) else {
                onProgress(Progress(operationId: id, bytesProcessed: 0, totalBytes: 0,
                                    done: true, error: "Cannot write: \(outputPath)",
                                    errorCode: "corruptStream"))
                return
            }
            defer { try? output.close() }

            let total = (try? FileManager.default
                .attributesOfItem(atPath: inputPath)[.size] as? Int64) ?? 0
            var position: Int64 = 0

            while true {
                if isCancelled(id) {
                    try? output.close()
                    try? FileManager.default.removeItem(atPath: outputPath)
                    onProgress(Progress(operationId: id, bytesProcessed: position,
                                        totalBytes: total ?? 0, done: true,
                                        error: "Cancelled", errorCode: "unknown"))
                    return
                }
                var data = input.readData(ofLength: chunk)
                if data.isEmpty { break }
                adapter.transform(&data, filePosition: position)
                output.write(data)
                position += Int64(data.count)
                onProgress(Progress(operationId: id, bytesProcessed: position,
                                    totalBytes: total ?? 0, done: false))
            }
            onProgress(Progress(operationId: id, bytesProcessed: position,
                                totalBytes: total ?? 0, done: true))
            clear(id)
        }
        return id
    }

    func cancel(_ id: String) {
        lock.lock(); defer { lock.unlock() }
        cancelled.insert(id)
    }

    private func isCancelled(_ id: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return cancelled.contains(id)
    }

    private func clear(_ id: String) {
        lock.lock(); defer { lock.unlock() }
        cancelled.remove(id)
    }
}

File: android/src/main/kotlin/com/hulkenstein/secure_video_player/FileCryptor.kt
package com.hulkenstein.secure_video_player

import java.io.File
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Streams a file through a [CipherAdapter] on a background thread in 1 MB
 * chunks — constant memory for any file size. Encrypt and decrypt are the
 * same operation for all built-in schemes (XOR keystreams are involutions).
 */
object FileCryptor {

    private const val CHUNK = 1024 * 1024

    data class Progress(
        val operationId: String,
        val bytesProcessed: Long,
        val totalBytes: Long,
        val done: Boolean,
        val error: String? = null,
        val errorCode: String? = null,
    )

    private val executor = Executors.newSingleThreadExecutor { r ->
        Thread(r, "SecureVideoCryptor").apply { isDaemon = true }
    }
    private val cancelled = ConcurrentHashMap<String, AtomicBoolean>()

    fun start(
        inputPath: String,
        outputPath: String,
        adapter: CipherAdapter,
        onProgress: (Progress) -> Unit,
    ): String {
        val id = UUID.randomUUID().toString()
        val flag = AtomicBoolean(false)
        cancelled[id] = flag

        executor.execute {
            val output = File(outputPath)
            try {
                val input = File(inputPath)
                if (!input.exists()) {
                    onProgress(Progress(id, 0, 0, true,
                        "Input not found: $inputPath", "fileNotFound"))
                    return@execute
                }
                val total = input.length()
                var position = 0L
                val buffer = ByteArray(CHUNK)

                input.inputStream().use { ins ->
                    output.outputStream().use { outs ->
                        while (true) {
                            if (flag.get()) {
                                outs.close()
                                output.delete()
                                onProgress(Progress(id, position, total, true,
                                    "Cancelled", "unknown"))
                                return@execute
                            }
                            val read = ins.read(buffer)
                            if (read == -1) break
                            adapter.transform(buffer, 0, read, position)
                            outs.write(buffer, 0, read)
                            position += read
                            onProgress(Progress(id, position, total, false))
                        }
                    }
                }
                onProgress(Progress(id, position, total, true))
            } catch (e: Exception) {
                output.delete()
                onProgress(Progress(id, 0, 0, true,
                    e.message ?: e.javaClass.simpleName, "corruptStream"))
            } finally {
                cancelled.remove(id)
            }
        }
        return id
    }

    fun cancel(operationId: String) {
        cancelled[operationId]?.set(true)
    }
}
